Adds sample_weight option to estimators fit method

kota7 opened the pull request:
### Description

Adds `sample_weight` option to the `fit` method of estimators.
Aims to cover the followings:

- [x] StackingClassifier
- [x] StackingCVClassifier
- [x] EnsembleVoteClassifier
- [x] StackingRegressor
- [x] StackingCVRegressor



### Related issues or pull requests

Fixes #438 

<!--  
If applicable, please link related issues/pull request here. E.g.,   
Fixes #366
-->



### Pull Request Checklist

- [x] Added a note about the modification or contribution to the `./docs/sources/CHANGELOG.md` file (if applicable)
- [x] Added appropriate unit test functions in the `./mlxtend/*/tests` directories (if applicable)
- [x] Modify documentation in the corresponding Jupyter Notebook under `mlxtend/docs/sources/` (if applicable)
- [x] Ran `nosetests ./mlxtend -sv` and make sure that all unit tests pass (for small modifications, it might be sufficient to only run the specific test file, e.g., `nosetests ./mlxtend/classifier/tests/test_stacking_cv_classifier.py -sv`)
- [x] Checked for style issues by running `flake8 ./mlxtend`




<!--NOTE  
Due to the improved GitHub UI, the squashing of commits is no longer necessary.
Please DO NOT SQUASH commits since they help with keeping track of the changes during the discussion).
For more information and instructions, please see http://rasbt.github.io/mlxtend/contributing/  
-->

pep8speaks commented:
Hello @kota7! Thanks for submitting the PR.

 - In the file [`mlxtend/regressor/stacking_regression.py`](https://github.com/kota7/mlxtend/blob/7900a89f01f7663fdf18907f9a067e9f70594776/mlxtend/regressor/stacking_regression.py), following                are the PEP8 issues :
> [Line 69:80](https://github.com/kota7/mlxtend/blob/7900a89f01f7663fdf18907f9a067e9f70594776/mlxtend/regressor/stacking_regression.py#L69): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)


 - There are no PEP8 issues in the file [`mlxtend/regressor/tests/test_stacking_regression.py`](https://github.com/kota7/mlxtend/blob/7900a89f01f7663fdf18907f9a067e9f70594776/mlxtend/regressor/tests/test_stacking_regression.py) !

rasbt commented on mlxtend/regressor/stacking_regression.py:
```diff
@@ -111,6 +111,8 @@ def fit(self, X, y):
         y : array-like, shape = [n_samples] or [n_samples, n_targets]
             Target values.
+        sample_weight : array-like, shape = [n_samples], optional
+            Sample weights.
```
Could you additionally specify that these are used by both the level-1 and meta-regressors?

kota7 replied:
Sure.  And the meta regressor too.

rasbt replied:
Yeah, maybe as doc string sth like

Sample weights passed as `sample_weights` to each regressor in the `regressors` list as well as the `meta_regressor`.

The rest of the PR looks great btw! Thanks!

kota7 commented:
@rasbt This thread may not be the best place to talk about this, but when I update the test script `test_stacking_cv_regressor.py`, I found that generating another random vector make existing tests fail.

For example, below this part:
https://github.com/rasbt/mlxtend/blob/030c1b703ffd52aef7d6b4a6e003451edd02f4fb/mlxtend/regressor/tests/test_stacking_cv_regression.py#L22-L30

if you add another line:
```
w = np.random.random(40)
```
then tests start to fail.  This is perhaps because cross validation generators share random seed with numpy.
To see this, the below code generates two different results, due to the call for `np.random.random` after setting seed.

```
import numpy as np
from sklearn.model_selection import KFold

np.random.seed(1)
cv = KFold(2, shuffle=True)
print(list(cv.split([1,2,3,4,5,6])))

np.random.seed(1)
np.random.random(10)
cv = KFold(2, shuffle=True)
print(list(cv.split([1,2,3,4,5,6])))
```

As a side effect, when I some test raises exception, then the random seed changes, which makes other tests to fail.  Due to this, error report becomes confusing since tests that have no problem fail.

A possible workaround is to explicitly define cross validation object with specific random state at the top of script, and keep using it in tests.  Would you want me to do this, or do you have any thoughts?

rasbt commented:
Thanks for bringing this up. This is actually a bit of a messy unit test design. The random seed should either be set for each function individually or it should be using the `random_state` via scikit-learn, which is probably the better solution.

I think the reason why the StackingCVRegressor currently doesn't have a `random_state` parameter is because it uses the `sklearn.model_selection.check_cv` method, which doesn't support that either. 

So, one option would be resetting the random seed in each unit test seems to be the best option for now, but that might be a lot of work, unfortunately, because the unit tests might produce different results then like you mentioned.

Your suggestion

> A possible workaround is to explicitly define cross validation object with specific random state at the top of script, and keep using it in tests.

is probably better because in case we change the the StackingCVRegressor in future to have its own random_state, a fixed Kfold object wouldn't require changing all the unit tests again

kota7 commented:
@rasbt I think it is ready for review. Thanks.

rasbt commented:
This looks really nice overall. I am wondering though...,  don't the sklearn estimators accept `sample_weight=None`? This could simplify a lot of the if/else statements that are currently in there:

            regr.fit(X, y)		             if sample_weight is None:
                 regr.fit(X, y)
             else:
                 regr.fit(X, y, sample_weight=sample_weight)

kota7 commented:
As far as I know, `Lasso`, `MLPClassifier`, `KNerighborsClassifier` do not support it, hence raises exception if you give `sample_weight` to them.  This piece of unittest catches that:

https://github.com/rasbt/mlxtend/blob/b46079a21209220f92ee274f4498e120e9a1b762/mlxtend/regressor/tests/test_stacking_regression.py#L132-L146

That is, if you code just `regr.fit(X, y, sample_weight=sample_weight)`, `sample_weight=None` is passed to `Lasso`, which raises error.

rasbt commented:
Good point! Otherwise the PR seems fine and I'd be happy to merge. Or do you have any additions in mind?

rasbt commented:
Awesome! Thanks for this PR, really appreciate it!
